docstore/memdocstore/mem.go

```diff
@@ -419,4 +431,48 @@ func (c *collection) As(i interface{}) bool { return false }
 func (c *collection) ErrorAs(err error, i interface{}) bool { return false }
 
 // Close implements driver.Collection.Close.
-func (c *collection) Close() error { return nil }
+// If the collection was created with  Filename option, Close writes the collection's
+// documents to the file.
+func (c *collection) Close() error {
+	return saveDocs(c.opts.Filename, c.docs)
+}
+
+type mapOfDocs = map[interface{}]map[string]interface{}
+
+// Read a map from the filename if is is not empty and the file exists.
+// Otherwise return an empty (not nil) map.
+func loadDocs(filename string) (mapOfDocs, error) {
+	if filename == "" {
+		return mapOfDocs{}, nil
+	}
+	f, err := os.Open(filename)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return nil, err
+		}
+		// If the file doesn't exist, return an empty map without error.
+		return mapOfDocs{}, nil
+	}
+	defer f.Close()
+	var m mapOfDocs
+	if err := gob.NewDecoder(f).Decode(&m); err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+// saveDocs saves m to filename if filename is not empty.
+func saveDocs(filename string, m mapOfDocs) error {
+	if filename == "" {
+		return nil
+	}
+	f, err := os.Create(filename)
+	if err != nil {
+		return err
+	}
+	if err := gob.NewEncoder(f).Encode(m); err != nil {
+		_ = f.Close()
+		return err
+	}
+	return f.Close()
```

Comment:
Is this error important to return? Maybe it can just be simplified as

```go
defer f.Close()
return gob.NewEncoder(f).Encode(m)
```

Reply:
When writing, `Close` might return an important error (because it may be flushing a write buffer, for example). You should also check errors from `Close` when writing.